pkg/controller/controller_binding.go

```diff
@@ -348,39 +351,64 @@ func isPlanBindable(serviceClass *v1alpha1.ServiceClass, plan *v1alpha1.ServiceP
 }
 
 func (c *controller) injectBinding(binding *v1alpha1.Binding, credentials map[string]interface{}) error {
-	glog.V(5).Infof("Creating Secret %v/%v", binding.Namespace, binding.Spec.SecretName)
-	secret := &v1.Secret{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      binding.Spec.SecretName,
-			Namespace: binding.Namespace,
-		},
-		Data: make(map[string][]byte),
-	}
+	glog.V(5).Infof("Creating/updating Secret %v/%v", binding.Namespace, binding.Spec.SecretName)
 
+	secretData := make(map[string][]byte)
 	for k, v := range credentials {
 		var err error
-		secret.Data[k], err = serialize(v)
+		secretData[k], err = serialize(v)
 		if err != nil {
+			// Terminal error
+			// TODO mark as terminal error once we have the terminal condition
 			return fmt.Errorf("Unable to serialize credential value %q: %v; %s",
 				k, v, err)
 		}
 	}
 
-	found := false
-
-	_, err := c.kubeClient.Core().Secrets(binding.Namespace).Get(binding.Spec.SecretName, metav1.GetOptions{})
+	// Creating/updating the Secret
+	secretClient := c.kubeClient.Core().Secrets(binding.Namespace)
+	existingSecret, err := secretClient.Get(binding.Spec.SecretName, metav1.GetOptions{})
 	if err == nil {
-		found = true
-	}
-
-	if found {
-		_, err = c.kubeClient.Core().Secrets(binding.Namespace).Update(secret)
+		// Update existing secret
+		if !IsControlledBy(existingSecret, binding) {
+			controllerRef := GetControllerOf(existingSecret)
+			// TODO mark as terminal error once we have the terminal condition
+			return fmt.Errorf("Secret '%s' is not owned by Binding, controllerRef: %v",
+				existingSecret.Name, controllerRef)
+		}
+		existingSecret.Data = secretData
+		_, err = secretClient.Update(existingSecret)
+		if apierrors.IsConflict(err) {
+			// Conflicting update detected, try again later
+			return fmt.Errorf("Conflicting Secret '%s' update detected", existingSecret.Name)
+		}
 	} else {
-		_, err = c.kubeClient.Core().Secrets(binding.Namespace).Create(secret)
+		if !apierrors.IsNotFound(err) {
+			// Terminal error
+			// TODO mark as terminal error once we have the terminal condition
+			return err
+		}
+		// Create new secret
+		secret := &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      binding.Spec.SecretName,
+				Namespace: binding.Namespace,
+				OwnerReferences: []metav1.OwnerReference{
+					*NewControllerRef(binding, bindingControllerKind),
+				},
+			},
+			Data: secretData,
+		}
+		_, err = secretClient.Create(secret)
+		if apierrors.IsAlreadyExists(err) {
```

Comment:
If we get some other kind of error is this treating it as a success since it doesn't stop processing?